File: Chapter 02/Python/section_2_3_3_testability/after/real_database_connection.py
from .database_connection import IDatabaseConnection

class RealDatabaseConnection(IDatabaseConnection):
    """
    The 'real' production implementation of the IDatabaseConnection.
    """

    def __init__(self, connection_string: str):
        self._connection_string = connection_string
        print(f"\n  [DB] Connecting to... {self._connection_string}")

    def get_data(self, query: str) -> list[str]:
        print(f"  [DB] Executing query: {query}")
        return ["real_data_row1", "real_data_row2"]

File: Chapter 02/C#/Section_2_3_3_Testability/After/RealDatabaseConnection.cs
using System;
using System.Collections.Generic;

namespace Chapter02.Testability.After;

/// <summary>
/// This is the "real" or "production" implementation of the `IDatabaseConnection` interface.
/// It would contain the actual logic to connect to and query a live database.
/// In a real-world application, this class would use a library like Dapper or Entity Framework Core
/// to interact with a SQL Server, PostgreSQL, or other database.
/// </summary>
public class RealDatabaseConnection : IDatabaseConnection
{
    private readonly string _connectionString;

    /// <summary>
    /// Initializes the real database connection.
    /// </summary>
    /// <param name="connectionString">The database connection string.</param>
    public RealDatabaseConnection(string connectionString)
    {
        _connectionString = connectionString;
        // In a real application, this is where the connection would be established.
        Console.WriteLine($"\n  [DB] Connecting to... {_connectionString}");
    }

    /// <summary>
    /// Fetches data from the live database.
    /// </summary>
    /// <param name="query">The query to execute.</param>
    /// <returns>A list of data rows from the real database.</returns>
    public List<string> GetData(string query)
    {
        // For demonstration purposes, we're just returning hardcoded data.
        // A real implementation would execute the query against the database
        // and return the results.
        Console.WriteLine($"  [DB] Executing query: {query}");
        return new List<string> { "real_data_row1", "real_data_row2" };
    }
}